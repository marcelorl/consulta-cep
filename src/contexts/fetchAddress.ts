import React from 'react'

import { fetchAddressByZipCode, fetchMapCoordinates } from '../services/address'

export default async (dispatch: any, zipCode: string) => {
  try {
    const addressData = await fetchAddressByZipCode(zipCode)
    const mapsCoordinates = await fetchMapCoordinates(zipCode)

    return dispatch({
      type: 'FETCH_ADDRESS',
      payload: {...addressData, mapsCoordinates}
    })
  } catch (e) {
    return dispatch({
      type: 'FETCH_ADDRESS_FAIL'
    })
  }
}
