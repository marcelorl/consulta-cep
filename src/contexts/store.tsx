import React from 'react'

import {
  IState,
  IContextProps,
  Actions
} from '../models'
import { fetchAddressByZipCode, fetchMapCoordinates } from '../services/address'

export const Store = React.createContext({} as IContextProps)

const initialState: IState = {
  address: {},
  status: ''
}

const reducer = (state: IState, action: Actions) => {
  switch (action.type) {
    case 'FETCH_ADDRESS':
      return { ...state, address: action.payload, status: 'LOADED' }
    case 'FETCH_ADDRESS_FAIL':
      return { ...state, address: {}, status: 'ERROR' }
    case 'ADDRESS_RESET': {
      return { ...state, address: {}, status: '' }
    }
    default:
      return state
  }
}

export const useAddress = () => {
  const [ state, dispatch ] = React.useReducer(reducer, initialState)

  const fetchAddress = async (zipCode: string) => {
    try {
      const addressData = await fetchAddressByZipCode(zipCode)
      const mapsCoordinates = await fetchMapCoordinates(addressData)

      return dispatch({
        type: 'FETCH_ADDRESS',
        payload: {...addressData, mapsCoordinates}
      })
    } catch (e) {
      return dispatch({
        type: 'FETCH_ADDRESS_FAIL'
      })
    }
  }

  const resetState = () =>
    dispatch({
      type: 'ADDRESS_RESET'
    })

  return [ state, { fetchAddress, resetState } ]
}

export const StoreProvider = (props: any) => {
  const [ state, action ] = useAddress()

  return (
    <Store.Provider value={{ state, action }}>
      {props.children}
    </Store.Provider>
  )
}
