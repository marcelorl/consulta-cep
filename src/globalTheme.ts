import { DefaultTheme } from 'styled-components'

export const globalTheme: DefaultTheme = {
  colors: {
    darkgray: '#999999',
    gainsboro: '#e0ded9',
    lightblue: '#aadaff',
    white: '#ffffff',
  }
}
