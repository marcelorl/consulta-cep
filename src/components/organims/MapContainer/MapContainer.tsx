import React, { Fragment } from 'react'
import styled from 'styled-components'

import { Store } from '../../../contexts/store'
import Map from '../../molecules/Map'
import closeSvg from './images/close.svg'

const MapContainerWrapper = styled.div`
  border: 2px solid ${props => props.theme.colors.gainsboro};
  box-sizing: border-box;
  margin-top: 1rem;
  padding: 3rem;
  position: relative;
`

const AddressWrapper = styled.div`
  margin-bottom: 2rem;

  > * {
    margin-bottom: 1rem;  
  }
`

const Street = styled.h2`
  font-size: 1.5rem;
  font-weight: bold;
`

const CloseButton = styled.button`
  background: transparent;
  border: none;
  cursor: pointer;
  position: absolute;
  right: 1rem;
  top: 1rem;
  
  img {
    height: 1rem;
    width: 1rem;    
  }
`

const MapContainer = () => {
  const { action, state } = React.useContext(Store)

  const onClickCloseButton = () => {
    return action.resetState()
  }

  const shouldShowMap = () => {
    if (state.status === 'ERROR') return <div>CEP nÃ£o encontrado!</div>

    return (
      <Fragment>
        <AddressWrapper>
          <Street>{state.address.logradouro}</Street>
          <p>{state.address.bairro}</p>
          <p>{state.address.localidade} - {state.address.uf}</p>
          <p>{state.address.cep}</p>
        </AddressWrapper>

        <Map coordinates={state.address.mapsCoordinates} />
      </Fragment>
    )
  }

  if (!['LOADED', 'ERROR'].includes(state.status)) return null

  return (
    <MapContainerWrapper >
      <CloseButton type="button" onClick={onClickCloseButton}>
        <img src={closeSvg} alt="close"/>
      </CloseButton>

      {shouldShowMap()}
    </MapContainerWrapper>
  )
}


export default MapContainer
