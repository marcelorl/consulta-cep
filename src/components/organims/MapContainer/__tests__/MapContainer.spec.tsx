import React from 'react'
import { ThemeProvider } from 'styled-components'
import { cleanup, fireEvent, render } from 'react-testing-library'

import MapContainer from '../MapContainer'
import { Store, useAddress } from '../../../../contexts/store'
import { globalTheme } from '../../../../globalTheme'

afterEach(cleanup)

describe('#MapContainer', () => {
  it('should not call fetch, because zipcode is required', () => {
    const StoreProvider = (props: any) => {
      const [ _, action ] = useAddress()
      const state = { address: {}, status: 'ERROR' }

      return (
        <Store.Provider value={{ state, action }}>
          {props.children}
        </Store.Provider>
      )
    }

    const { container, getByText } = render(
      <StoreProvider>
        <ThemeProvider theme={globalTheme}>
          <MapContainer />
        </ThemeProvider>
      </StoreProvider>
    )

    expect(getByText('CEP n√£o encontrado!')).not.toBeNull()
    expect(container).toMatchSnapshot()
  })
})


