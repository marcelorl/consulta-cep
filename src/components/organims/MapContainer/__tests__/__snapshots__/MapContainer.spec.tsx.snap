// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#MapContainer should not call fetch, because zipcode is required 1`] = `
.c0 {
  border: 2px solid #e0ded9;
  box-sizing: border-box;
  margin-top: 1rem;
  padding: 3rem;
  position: relative;
}

.c1 {
  background: transparent;
  border: none;
  cursor: pointer;
  position: absolute;
  right: 1rem;
  top: 1rem;
}

.c1 img {
  height: 1rem;
  width: 1rem;
}

<div>
  <div
    class="c0"
  >
    <button
      class="c1"
      type="button"
    >
      <img
        alt="close"
        src="close.svg"
      />
    </button>
    <div>
      CEP n√£o encontrado!
    </div>
  </div>
</div>
`;
